src/java/maven2/handlers/maven_artifacts.rs

```diff
@@ -0,0 +1,165 @@
+/*
+   Copyright 2021 JFrog Ltd
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+use crate::artifact_service::handlers::get_artifact;
+use crate::artifact_service::storage::ArtifactStorage;
+use crate::docker::error_util::{RegistryError, RegistryErrorCode};
+use crate::network::client::Client;
+use crate::transparency_log::log::TransparencyLog;
+use anyhow::anyhow;
+
+use log::debug;
+use warp::{http::StatusCode, Rejection, Reply};
+
+pub async fn handle_get_maven_artifact(
+    transparency_log: TransparencyLog,
+    p2p_client: Client,
+    artifact_storage: ArtifactStorage,
+    full_path: String,
+) -> Result<impl Reply, Rejection> {
+    debug!("Requesting maven artifact: {}", full_path);
+    let namespace_specific_id = get_namespace_specific_id(&full_path).map_err(|_| {
+        warp::reject::custom(RegistryError {
+            code: RegistryErrorCode::BlobUnknown,
+        })
+    })?;
+
+    // request artifact
+    debug!("Requesting artifact for id {}", &namespace_specific_id);
+    let blob_content = get_artifact(
+        transparency_log,
+        p2p_client,
+        &artifact_storage,
+        &namespace_specific_id,
+    )
+    .await
+    .map_err(|err| {
+        debug!("Error retrieving artifact: {:?}", err);
+        warp::reject::custom(RegistryError {
+            code: RegistryErrorCode::BlobUnknown,
+        })
+    })?;
+
+    Ok(warp::http::response::Builder::new()
+        .header("Content-Type", "application/octet-stream")
+        .status(StatusCode::OK)
+        .body(blob_content)
+        .unwrap())
+}
+
+fn get_namespace_specific_id(full_path: &str) -> Result<String, anyhow::Error> {
+    // maven coordinates like "com.company:test:1.0" will produce a request
+    // like: "GET /maven2/com/company/test/1.0/test-1.0.jar"
+
+    // split, and remove first two strings: "" and "maven2":
+    let mut pieces: Vec<&str> = full_path.split('/').skip(2).collect();
+    if pieces.len() < 4 {
+        return Err(anyhow!(format!("Error, invalid full path: {}", full_path)));
```

Comment:
This can be simplified with the `anyhow::bail` macro:
```suggestion
        bail!(format!("Error, invalid full path: {}", full_path));
```

Perhaps you can add a test that executes this code branch?